Update alexa integration with new metadata syntax

jsetton opened the pull request:
Related to openhab/openhab-alexa#394.

This change should be merged after the new metadata syntax is released to the production environment.

Since the new syntax is backward compatible, users of previous GUI versions will still be able to use the old interface.

jsetton commented:
@ghys could you take a look at this one when you have a chance? It's pretty much ready for review. They will be some minor data input changes based on the beta testing feedback and bug fixes. I just want to make sure that we can merge this one once we move the Alexa metadata changes to the live skill.

ghys commented on bundles/org.openhab.ui/web/src/components/item/metadata/item-metadata-alexa.vue:
```diff
@@ -1,35 +1,65 @@
+    <f7-list v-if="ready">
+      <f7-list-item
+        :key="classSelectKey"
+        :title="'Alexa Device Type' + (itemType !== 'Group' ? !multiple ? '/Attribute' : '/Attributes' : '')"
```
Maybe add some parentheses here in the inner ternary to make it easier to read?

jsetton replied:
Done

ghys commented on bundles/org.openhab.ui/web/src/assets/definitions/metadata/alexa/helpers.js:
```diff
@@ -0,0 +1,63 @@
+}
+
+export const link = (title, anchor) => {
+  const link = `https://www.openhab.org/link/alexa#${anchor || title.replace(/[. ]/g, '-').toLowerCase()}`
```
Beware that before the next stable version is released, up-to-date documentation is found at https://next.openhab.org, not www.openhab.org.
You can determine if you're running a release build with code like this:
```js
this.$store.state.runtimeInfo.buildString === 'Release Build' ? 'www' : 'next'
```
(but only in the context of a Vue component where `this` is defined), outside that context I'm not sure, maybe importing the store would work?

jsetton replied:
I added support for the different documentation links per your recommendation. For outside that context, I am using a placeholder `%DOC_URL%` that gets replaced once back in context.

ghys commented on bundles/org.openhab.ui/web/src/components/item/metadata/item-metadata-alexa.vue:
```diff
@@ -1,35 +1,65 @@
+    <div v-if="ready">
       <config-sheet :parameterGroups="[]" :parameters="parameters" :configuration="metadata.config" />
     </div>
+    <f7-block class="padding-top no-padding no-margin" v-if="itemType === 'Group' && classes.length">
```
Please add a `margin-left` class to make it look better when the screen is narrow:

![image](https://user-images.githubusercontent.com/2004147/132695766-40b22725-50fd-4434-be04-883bbec81ea0.png)

jsetton replied:
Done. I added `padding-left` class to the `f7-block-title` and `f7-block-footer` tags part of that block.

ghys commented on bundles/org.openhab.ui/web/src/components/item/metadata/item-metadata-alexa.vue:
```diff
@@ -47,43 +77,125 @@ export default {
+    },
+    link () {
+      return 'https://www.openhab.org/link/alexa#'.concat(
+        this.itemType === 'Group'
+          ? 'group-endpoint'
+          : this.classes.length === 0 || !this.classesDefs.includes(this.classes[0])
+            ? this.item.groups.length
+              ? 'group-endpoint'
+              : 'single-endpoint'
+            : this.classes[0].indexOf('.') === -1
+              ? this.classes[0] === 'Scene' || this.classes[0] === 'Activity'
+                ? this.classes[0].toLowerCase()
+                : 'device-types'
+              : this.classes[0].split('.')[1].toLowerCase()
```
This could surely be made more readable with some regular "if" statements :)

jsetton replied:
Done including the different documentation link support as mentioned above.

ghys reviewed (commented):
Thanks for the huge update @jsetton - I only have a few minor comments, apart from those the code looks good, on the functionality itself I'll trust you as I don't use Alexa or follow the latest developments.

jsetton commented:
@ghys sorry for the delay. I made the changes based on your review comments. Let me know what you think.

ghys commented:
@jsetton thanks - I have no more comments.
Remove the draft status to have it merged when the backend side is ready!

jsetton commented:
@ghys I decided to change how cross-device connections are defined with the network capabilities using group relationships in the end. This means that I reverted the items query to just pulling specific group items. So that should limit heavy requests.

As far as your last comment, I have moved the API requests from `created` to `mounted` Vue event. I also added a `f7-preloader` while `this.ready` is false in the event a specific item is a member of a high number of groups.

Hopefully, you are fine with the last change. This is the last functional change that I will be making on this PR.

ghys commented on bundles/org.openhab.ui/web/src/components/item/metadata/item-metadata-alexa.vue:
```diff
@@ -91,15 +91,14 @@ export default {
-        ? this.item.groupNames.map((groupName) => this.$oh.api.get(`/rest/items/${groupName}?metadata=alexa`))
-        : [])
-    ]).then(([regional, ...groups]) => {
+      this.$oh.api.get('/rest/items?metadata=alexa&fields=name,label,type,groupNames,groupType,metadata'),
```
Yes, my bad, obviously it's no page so `mounted` is the right place to put it!

ghys commented:
These changes are fine by me!

jsetton commented:
@ghys This PR is ready to be merged. Relevant changes have been merged and deployed to the production environment.

digitaldan commented:
Looking forward to seeing this merged!   @jsetton has our next release of the Alexa skill ready to go.

ghys reviewed (approved):
Thank you for this extensive overhaul! - since I don't use Alexa I will trust you on the functional side, but the UI seems to work at it should, and the code lgtm.
